gapic/src/main/com/google/gapid/views/VirtualTreeViewer.java

```diff
@@ -0,0 +1,143 @@
+   * implement to be notified about {@link TreeItem}s being made visible and hidden
+   * usually by scrolling.
+   */
+  public interface VisibilityListener {
+    /**
+     * Called when the {@link TreeItem} is made visible.
+     */
+    void OnShow(TreeItem item);
+    /**
+     * Called when the {@link TreeItem} is made invisible.
+     */
+    void OnHide(TreeItem item);
+  }
+
+  private Set<TreeItem> visible = new HashSet<>();
+
+  public VirtualTreeViewer(Tree tree) {
+    super(tree);
+    tree.addPaintListener((e) -> updateVisibility());
+  }
+
+  private void updateVisibility() {
+    List<VisibilityListener> listeners = getListeners();
+    if (listeners.size() == 0) {
+      return;
+    }
+
+    Set<TreeItem> seen = calcVisibleItems();
+    if (seen == null) {
+      return; // No reliable data.
+    }
+    for (TreeItem item : visible) {
+      if (!seen.contains(item)) {
+        for (VisibilityListener listener : listeners) {
+          listener.OnHide(item);
+        }
+      }
+    }
+    for (TreeItem item : seen) {
+      if (!visible.contains(item)) {
+        for (VisibilityListener listener : listeners) {
+          listener.OnShow(item);
+        }
+      }
+    }
+    visible = seen;
+  }
+
+  private List<VisibilityListener> getListeners() {
+    List<VisibilityListener> listeners = new ArrayList<>();
+    IContentProvider contentProvider = getContentProvider();
+    if (contentProvider instanceof VisibilityListener) {
+      listeners.add((VisibilityListener)contentProvider);
+    }
+    IBaseLabelProvider labelProvider = getLabelProvider();
+    if (labelProvider instanceof VisibilityListener) {
+      listeners.add((VisibilityListener)labelProvider);
+    }
+    return listeners;
+  }
+
+  private TreeItem nextItem(TreeItem item) {
+    if (item.getExpanded()) {
+      // Descend into children.
+      int children = TreeHelper.itemCount(item);
+      if (children > 0) {
+        return TreeHelper.getItem(item, 0);
+      }
+    }
+    // Ascend ancestors looking for siblings.
+    Object parent = TreeHelper.getParent(item);
+    while (parent != null) {
+      int index = TreeHelper.indexOf(parent, item);
+      int siblings = TreeHelper.itemCount(parent);
+      if (index+1 < siblings) {
+        return TreeHelper.getItem(parent,index + 1);
+      }
+      if (!(parent instanceof TreeItem)) {
+        break;
+      }
+      item = (TreeItem)parent;
+      parent = TreeHelper.getParent(parent);
+    }
+    return null;
+  }
+
+  private Set<TreeItem> calcVisibleItems() {
+    Set<TreeItem> visible = new HashSet<>();
+    Tree tree = VirtualTreeViewer.this.getTree();
+    Rectangle rect = tree.getClientArea();
+    if (tree.getTopItem() == null && tree.getItemCount() != 0) {
+      // Work around bug where getTopItem() returns null when scrolling
+      // up past the top item (elastic scroll).
+      return null;
+    }
+    for (TreeItem item = tree.getTopItem(); item != null; item = nextItem(item)) {
```

Comment:
Doing this iteratively seems to make it more complicated. I think most of the complexity of nextItem could be avoided if this were done recursively. (along with no more need for the tree helper class)